add new 'swap_sources' action and others

seb26 opened the pull request:
**New functionality**:

* added new action 'swap sources' (*Swap the sources of two specified destinations*) - saves the current source for destinations A and B, and then routes them to the opposite destination.

Example usage - for swapping the position of two video sources on a multiviewer.

* added new feedback 'source_match' - used on a 'Source' button, and illuminates if user selects a 'Destination' button and that destination matches the button's source

When used in combination with 'active_destination' on a 'Destination' button, this mimics the behaviour of the physical KUMO CP panels, that illuminate the currently routed source when you choose to cycle through the 'Destination' buttons without making any actual routing change.

----

**Changes to existing functions**:

* actions - allowed for multiple API calls to be made within the same action - upon execute, it will loop and send each one

* action 'route' – changed label wording from "output" to "destination" to match all other terminology use

* actions 'destination' & 'source' – changed to use a Dropdown menu to indicate In/Out names from API, to aid identifying which in/out compared to just a number field - number can still be inputted manually by typing it

* feedbacks - changed to use a Dropdown menu as above

onfire4g05 commented:
Great addition! I've been looking over it and wondering if maybe we could just allow variable usage for these instead and it may be better/accomplish the same result? The only challenge you then have is you would have to use a custom variable to store the first value. So something like this:

<img width="562" alt="Screen Shot 2022-09-19 at 8 31 06 AM" src="https://user-images.githubusercontent.com/991168/191028946-1ebabae9-756a-4bba-977b-759c5e3bb8da.png">

I've also considered marking the old source / destination logic as "legacy" and recommending people to use the route instead, since I would think that's usually what people would want, but maybe I'm completely wrong there or have the wrong terminology (ie, legacy != one day it will be removed)? My thought behind that was it doesn't do anything with the device until you actually change the source -- ie, when companion is restarted the old values won't be there -- and you'd be better able to use feedbacks for when X routes to Y and everything is then stateful.

I added the variable usage concept to my [companion3-upgrade branch](https://github.com/onfire4g05/companion-module-aja-kumo/blob/companion3-upgrade/index.js); I only realized today I forgot to finish the most important method in that though... `actionCall`. I'll get that resolved soon.

seb26 commented:
Thanks @onfire4g05! 

For swapping two sources: I actually considered variables at the beginning and good to see there was an easy way to enable that. Can see lots of uses for variables in that manner.

However, would you accept a new action 'swap_sources' anyway, if I quickly adapted it to use your newer actionCall() function, instead of the `cmds` array that I wrote in [4fc3147](https://github.com/bitfocus/companion-module-aja-kumo/pull/12/commits/4fc3147c14472af6fccf27fc2e8c139916c668a2)? An action for that purpose is a lot easier to use than 3 actions and a custom variable. It would actually be quite tedious to do it that way and can't see too much harm from an action of this kind.

For "old source / destination logic as "legacy"":

To give you my thoughts, I would suggest not to remove this functionality. The behaviour people are looking for is that which is present in the AJA KUMO CP physical control panels. Selecting a destination button illuminates both that button, and simultaneously illuminates the source that is routed to it. It becomes the primary visual means of determining which source is routed to which destination, and on the KUMO CP it is expected from a usage perspective that you can press on a destination and see its source without making a change at all. I.e. you can press on destinations 1, 2, 3 and keep going through them and the different 'Sources' will light up as you go. It is necessary because on the AJA routers anyway there is no onscreen display or any other means to 'see' the routing.

So for Companion, an action that temporarily saves a 'Destination' as a target, plus an action that follows where a source is selected to complete the Source -> Destination route, are necessary to mimic that behaviour. It will not always be known to the user, upon 1 single button press, the exact Source & Destination they wish to route to.

When I create a matrix of buttons in Companion, e.g. 16 buttons for Sources and 16 buttons for Destinations, I am able to put `source_match` feedback on each Source button and `active_destination` feedback on each Destination button, and the behaviour matches the KUMO CP basically identically. I wasn't able to do this prior to writing `source_match` in this pull request but now can. This is what issue #5 is referring to but I don't believe it was properly resolved at the time that issue was closed. 

As a final note I agree that the state of a selected destination would only be maintained during that instance of companion, but having it return to a default value after a program reboot would not be counterproductive. The destination selected by the user is only relevant to them at the time they take the subsequent step to pick a source.

seb26 commented:
Thanks for that!

I have fixed the tabs and used .forEach() in this fork of mine. And then on your v3 I adapted the 'swap_sources' action and 'source_match' feedback.

Let me know if anything else needs fixing.

onfire4g05 commented on feedbacks.js:
```diff
@@ -29,16 +30,38 @@ exports.initFeedbacks = function() {
+			label: 'Source number',
+			id: 'source',
+			default: 1,
+            choices: this.getNameList('src')
```
one more spacing issue

seb26 commented:
Thanks @onfire4g05 for your support! I'll post it to the channel now.

josephdadams commented:
Should this be merged, or closed?

seb26 commented:
So an update, I reviewed the code currently in companion-module-aja-kumo 2.0.2 and it looks like my original contributions have been adapted! Thank you @onfire4g05, I imagine you adapted it at some point?

I have also tested it with my Companion button layout in Companion 3.0.0rc2 and the Actions are functioning fully.

There is one feedback (`source_match`) which does not show up in the list of available Feedbacks. To my buttons that had it already added, the feedback shows up as "undefined".

I inspected the feedback and it seems laid out fine and no console errors. For some reason, this Feedback (which I'm using to just make the button go red) actually still works despite appearing as "undefined". However, I deleted it from one button and now I cannot add it back again because it is not in the list.

I observe this happening in just plain 2.0.2, but also in my patch-2 https://github.com/seb26/companion-module-aja-kumo/commit/1b8cf2ae7f0e88e037b8a3e4c2c72cae1bebe07f. In this patch I created a dummy function that just returns true or false. This dummy function also does not show up in the list of Feedbacks anywhere. 

Could anyone help explain why this is occurring?

Also in that commit @onfire4g05, I have expanded the HELP and adapted the labelling to explain the saved destination functionality within this module. I just straight up explained it as "Pre-select a destination and Companion will remember it". This is essentially how it works, when making route commands using 2 stage-button presses like I talked about in https://github.com/bitfocus/companion-module-aja-kumo/pull/12#issuecomment-1252164963. Therefore I removed the Legacy wording as well, since it is not an older legacy action competing with the `route` action (being a 1 stage action, where the source & dest need to be specified at the time of action), it's an alternative, so both of them need to co-exist.

seb26 commented:
Here is a screenshot of what I'm witnessing with the Feedback leftover from my Companion config in 2.4.1, that seems to function correctly (???) but if it is deleted, it cannot be added again.

<img width="540" alt="image" src="https://github.com/bitfocus/companion-module-aja-kumo/assets/931319/ff969c04-6630-4596-a061-dea3f0cbb890">

It should say kumo: source_match.

josephdadams commented:
It’s because that feedback does not have a “name” property. It’s still “label”. This was changed to name in v3

seb26 commented:
OK, this PR is in good shape now!

Summary of changes:
- No actual code changes for Swap Sources, feedbacks, etc - these were already added.
- Changed labelling of feedback/actions to reflect the 'Pre-selection' functionality.
- Changed labelling of Salvo to "Take" which is the terminology used by AJA and then (apply) in brackets to further clarify
- HELP expansion.

Also can discard my fix for parsedResponse, I must have had 2.0.0 in use at the time I encountered that and not 2.0.2.

onfire4g05 commented:
@seb26 Fix some of the space-tabs for actual tabs and combine multiple feedback checks into one (`this.checkFeedbacks('x', 'y')`). Also, since `destination_match` is the only feedback that's actually from the Kumo, let's change the description for that to hopefully make it more clear:

`When routing on the device changes to a specific source and destination.`

After that, I think this is good to merge.

seb26 commented:
Thanks for the feedback @onfire4g05 , done above.
